hypersync-client/src/from_arrow.rs

```diff
@@ -11,49 +12,105 @@ use crate::{
 /// Used to do ArrowBatch-Native Rust type conversions while consuming the input value.
 pub trait FromArrow: Sized {
     /// Converts to the Vector type from the ArrowBatch type.
-    fn from_arrow(batch: &ArrowBatch) -> Vec<Self>;
+    /// B is the type of binary array used to downcast since we use both binary views and
+    /// binary arrays
+    fn from_arrow_bin_array<B: List<Item = [u8]> + 'static, U: List<Item = str> + 'static>(
+        batch: &ArrowBatch,
+    ) -> Vec<Self>;
+
+    /// Default implementation that uses the binary array type.
+    /// IPC is written with regular binary arrays.
+    /// Named 'from_arrow' for backwards compatibility.
+    fn from_arrow(batch: &ArrowBatch) -> Vec<Self> {
+        Self::from_arrow_bin_array::<BinaryArray<i32>, Utf8Array<i32>>(batch)
+    }
+
+    /// An additional method that uses the binary view array type.
+    /// This is to be able to reuse the trait server side where Binary Views are
+    /// used instead.
+    fn from_arrow_bin_view_array(batch: &ArrowBatch) -> Vec<Self> {
+        Self::from_arrow_bin_array::<BinaryViewArray, Utf8ViewArray>(batch)
+    }
+}
+
+/// A simple trait to compose binary array types that need to be
+/// accessed by index
+pub trait List {
+    type Item: ?Sized;
+
+    fn get_idx(&self, i: usize) -> Option<&Self::Item>;
+}
+
+impl List for BinaryArray<i32> {
+    type Item = [u8];
+
+    fn get_idx(&self, i: usize) -> Option<&[u8]> {
+        self.get(i)
+    }
+}
+impl List for BinaryViewArray {
+    type Item = [u8];
+
+    fn get_idx(&self, i: usize) -> Option<&[u8]> {
+        self.get(i)
+    }
+}
+
+impl List for Utf8Array<i32> {
+    type Item = str;
+
+    fn get_idx(&self, i: usize) -> Option<&str> {
+        self.get(i)
+    }
+}
+
+impl List for Utf8ViewArray {
+    type Item = str;
+
+    fn get_idx(&self, i: usize) -> Option<&str> {
+        self.get(i)
+    }
```

Comment:
I made it possible to access by index all these different types based on a trait rather than having to use the actual type.